qutip/solver/nm_mcsolve.py

```diff
@@ -359,12 +372,16 @@ class NonMarkovianMCSolver(MCSolver):
     }
 
     def __init__(
-        self, H, ops_and_rates, args=None, options=None,
+        self,
+        H: Qobj | QobjEvo,
+        ops_and_rates = Sequence[tuple[Qobj, float | Coefficient]],
+        *,
+        options: dict[str, Any] = None,
```

Comment:
I am unsure about removing the `args` parameter here. It seems reasonable to me that anywhere that the user is allowed to pass a pure python function to describe time dependence, they should also be able to pass `args`.

Also, the type hint for `ops_and_rates` doesn't correspond to the actual code. What we actually accept, currently, is `Sequence[tuple[Qobj, float | CoefficientLike]]`. Therefore, users are able to pass callables there, and I would expect `args` to be possible.

Note that for the constructor of MCSolver, the type hint for `H` is `QobjEvoLike`, but I think currently it only accepts `Qobj | QobjEvo` (because of [this issuper](https://github.com/qutip/qutip/blob/master/qutip/solver/mcsolve.py#L467)). Therefore, an `args` parameter is unneccessary. If the code is changed such that any `QobjEvoLike` is accepted, I would argue to add an `args` parameter also there.

Reply:
All `??Solver` do not take `QobjEvoLike`, only pre-made `QobjEvo` by design. Thank you for catching.

The `args` description was wrong. It says that it was used for `Arguments for time-dependent Hamiltonian and collapse operator terms.`. But it these operators should already be `QobjEvo` and do not use this `args`. 

`CoefficientLike` support array format, thus need the `tlist`, which does not belong there.  In the `??solve` functions, we accept anything an try to do our best to make sense of it, but it's messy... With the class interface, I try to be more strict and avoid these cases.

Reply:
Okay, that makes sense to me. I went to the changes and found a few more small inconsistencies, I think. Overall, very nice to have some type hints, thank you!